lib/Dialect/Seq/SeqOps.cpp

```diff
@@ -449,8 +449,26 @@ ParseResult FirRegOp::parse(OpAsmParser &parser, OperationState &result) {
   result.addTypes({ty});
 
   if (presetValue) {
-    result.addAttribute("preset",
-                        parser.getBuilder().getIntegerAttr(ty, *presetValue));
+    uint64_t width = 0;
+    if (hw::type_isa<seq::ClockType>(ty)) {
+      width = 1;
+    } else {
+      int64_t maybeWidth = hw::getBitWidth(ty);
+      if (maybeWidth < 0)
+        return parser.emitError(loc, "cannot preset register of unknown width");
+      width = maybeWidth;
+    }
+
+    APInt presetResult = presetValue->sextOrTrunc(width);
+    if (presetValue->getBitWidth() > width) {
+      if (presetResult.sext(presetValue->getBitWidth()) != *presetValue)
+        return parser.emitError(loc, "integer value too large");
+    }
```

Comment:
I think this logic may not be correct given that this is rejecting both of:

``` mlir
  %c = seq.firreg %x clock %clock preset 3 : i2
  %d = seq.firreg %x clock %clock preset 2 : i2
```

With:

```
/Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/Seq/round-trip.mlir:83:19: error: custom op 'seq.firreg' integer value too large
  %c = seq.firreg %x clock %clock preset 2 : i2
                  ^
/Users/schuylereldridge/repos/github.com/llvm/circt/test/Dialect/Seq/round-trip.mlir:83:19: error: custom op 'seq.firreg' integer value too large
  %d = seq.firreg %x clock %clock preset 3 : i2
                  ^
```

Can the PR add a test that presets that use all the available bits still work?

Can this test the same thing for a constant that is the full bit width of an aggregate?

Nit: the error message using `loc` is not placing the carat in the correct place. A user might be confused with the error message here thinking that `%x` was wrong.